packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -895,188 +895,253 @@ class TextPainter {
+    if (position == _previousCaretPosition) {
+      return _cachedCaretMetrics;
+    }
+    _previousCaretPosition = position;
+
+    if (isPreviousCodeUnitLineBreak && currentCodeUnit == null) {
+      final _CaretMetrics? downstream = _searchForCaretAnchor(paragraph, text, position, 1);
+      if (downstream != null) {
+        return _cachedCaretMetrics = downstream;
+      }
+      // Special case for the HTML renderer getBoxesForRange implementation.
+      // previousCodeUnit is a newline so the return value has to be non-null.
+      final _CaretMetrics? upstream = _searchForCaretAnchor(paragraph, text, TextPosition(offset: position.offset, affinity: TextAffinity.upstream), -1);
+      assert(upstream != null);
+      final double shiftY = upstream == null ? 0.0 : upstream.fullHeight + upstream.offset.dy;
+      return _cachedCaretMetrics = _fallbakCaretMetrics.shiftDown(shiftY);
+    }
+
+    final bool multiCodePointUpstream = previousCodeUnit == null // This implies offset == 0
+                                     || _isUtf16Surrogate(previousCodeUnit)
+                                     || currentCodeUnit == _zwjUtf16
+                                     || _isUnicodeDirectionality(previousCodeUnit);
+    final bool multiCodePointDownstream = currentCodeUnit != null
+                                       && (_isUtf16Surrogate(currentCodeUnit) || currentCodeUnit == _zwjUtf16 || _isUnicodeDirectionality(currentCodeUnit));
+    final int startSearchOffset = multiCodePointUpstream ? -2 : -1;
+    final int endSearchOffset = multiCodePointDownstream ? 2 : 1;
+    final _CaretMetrics? caretMetrics;
     switch (position.affinity) {
-      case TextAffinity.upstream: {
-        rect = _getRectFromUpstream(offset, caretPrototype) ?? _getRectFromDownstream(offset, caretPrototype);
+      case TextAffinity.upstream:
+        caretMetrics = _searchForCaretAnchor(paragraph, text, position, startSearchOffset)
+          ?? (searchOppositeAffinity ? _searchForCaretAnchor(paragraph, text, TextPosition(offset: position.offset), endSearchOffset) : null);
         break;
-      }
-      case TextAffinity.downstream: {
-        rect = _getRectFromDownstream(offset, caretPrototype) ??  _getRectFromUpstream(offset, caretPrototype);
+      case TextAffinity.downstream:
+        caretMetrics = _searchForCaretAnchor(paragraph, text, position, endSearchOffset)
+          ?? (searchOppositeAffinity ? _searchForCaretAnchor(paragraph, text, TextPosition(offset: position.offset, affinity: TextAffinity.upstream), startSearchOffset) : null);
         break;
-      }
     }
-    _caretMetrics = _CaretMetrics(
-      offset: rect != null ? Offset(rect.left, rect.top) : _emptyOffset,
-      fullHeight: rect != null ? rect.bottom - rect.top : null,
-    );
-
-    // Cache the input parameters to prevent repeat work later.
-    _previousCaretPosition = position;
-    _previousCaretPrototype = caretPrototype;
+    return _cachedCaretMetrics = caretMetrics;
```

Comment:
Why did you change this to both return and assign?

Reply:
caretPrototype doesn't affect the anchoring of the caret and it's not expensive at all to compute, there's no need to invalidate the cache over that.